:sparkles: Add feature gate marker support

wazery opened the pull request:
Implements `+kubebuilder:feature-gate=<gate-name>` marker that allows fields to be conditionally included in generated CRDs based on enabled feature gates.

- Add centralized featuregate package with evaluation and parsing logic
- Implement CRD feature gate markers for conditional field inclusion
- Add RBAC feature gate support with multiple gate evaluation
- Extend Webhook generator with feature gate capabilities
- Include comprehensive test coverage and golden output files

Addresses kubernetes-sigs/controller-tools#1238, and #600

wazery committed:
feat: Add comprehensive feature gate support for CRD, RBAC, and Webhook generators

- Add centralized featuregate package with evaluation and parsing logic
- Implement CRD feature gate markers for conditional field inclusion
- Add RBAC feature gate support with multiple gate evaluation
- Extend Webhook generator with feature gate capabilities
- Include comprehensive test coverage and golden output files

JoelSpeed reviewed (commented):
I'm keen to seen progress on this issue, thanks @wazery for starting the push forward.

Feature gating is complex when it comes to CRDs. I'd be keen to see if we can also work out a plan for how we will gate specific markers, is that something we can align on as a project before we move forward?

Specifically I've seen needs for feature gated different enum sets, different feature gated XValidations and different values for things like Min/Max items that are feature gated.

But perhaps we actually need to be able to support gating all of the various markers we support?

/ok-to-test

wazery commented:
> I'm keen to seen progress on this issue, thanks @wazery for starting the push forward.
> 
> Feature gating is complex when it comes to CRDs. I'd be keen to see if we can also work out a plan for how we will gate specific markers, is that something we can align on as a project before we move forward?
> 
> Specifically I've seen needs for feature gated different enum sets, different feature gated XValidations and different values for things like Min/Max items that are feature gated.
> 
> But perhaps we actually need to be able to support gating all of the various markers we support?
> 
> /ok-to-test

Thanks a lot Joel for your review, I truly appreciate it and learn much from it.

As you can see in the current implementation it's very initial, and allows only binary field inclusion so either the entire field is in or out. Yet we can enhance the implementation further to address your cases:

- Feature-gated enum values
- Feature-gated validation rules (XValidations)
- Feature-gated marker values (Min/MaxItems, etc.)
- Conditional marker application

For example:

## Option A: Inline Feature Gate Syntax
```
// Feature-gated enum values
// +kubebuilder:validation:Enum=stable;production
// +kubebuilder:validation:Enum=alpha;beta;experimental,featureGate=alpha

// Feature-gated validation rules  
// +kubebuilder:validation:XValidation=rule="self.size > 0"
// +kubebuilder:validation:XValidation=rule="self.advanced == true",featureGate=alpha

// Feature-gated min/max values
// +kubebuilder:validation:MinItems=1
// +kubebuilder:validation:MinItems=5,featureGate=strict
```

## Option B: Dedicated Feature Gate Markers

```
// +kubebuilder:validation:Enum=stable;production
// +kubebuilder:validation:Enum:featureGate=alpha,value=alpha;beta;experimental

// +kubebuilder:validation:XValidation=rule="self.size > 0"  
// +kubebuilder:validation:XValidation:featureGate=alpha,rule="self.advanced == true"
```

## Option C: Block-based Feature Gating
```
// +kubebuilder:feature-gate:alpha
// +kubebuilder:validation:Enum=alpha;beta;experimental  
// +kubebuilder:validation:XValidation=rule="self.advanced == true"
// +kubebuilder:feature-gate:end

// +kubebuilder:validation:Enum=stable;production  // always applied
```

I love to hear your thoughts on how do you think we have to move forward on this PR, and truly appreciated your help 🙇.

wazery commented on pkg/crd/markers/featuregate.go:
```diff
@@ -0,0 +1,53 @@
+		Category: "CRD feature gates",
+		DetailedHelp: markers.DetailedHelp{
+			Summary: "marks a field to be conditionally included based on feature gate enablement",
+			Details: "Fields marked with +kubebuilder:feature-gate will only be included in generated CRDs when the specified feature gate is enabled via --feature-gates flag.",
```
you are right, I just fixed it, thanks for spotting it.

JoelSpeed commented on pkg/crd/testdata/featuregates/types.go:
```diff
@@ -0,0 +1,74 @@
+	// Standard field - always included
+	Name string `json:"name"`
+
+	// Alpha-gated field - only included when alpha gate is enabled
+	// +kubebuilder:feature-gate=alpha
+	AlphaFeature *string `json:"alphaFeature,omitempty"`
+
+	// Beta-gated field - only included when beta gate is enabled  
+	// +kubebuilder:feature-gate=beta
+	BetaFeature *string `json:"betaFeature,omitempty"`
```
What would happen if I wanted one of these fields to be on when either gate was enabled, and then another field only when both were enabled?

I can see the `|` and `&` logic in the rbac, do we have that for the fields/markers as well?

JoelSpeed commented on pkg/rbac/advanced_feature_gates_test.go:
```diff
@@ -0,0 +1,143 @@
+		{name: "single gate", expression: "alpha", shouldError: false},
+		{name: "OR expression", expression: "alpha|beta", shouldError: false},
+		{name: "AND expression", expression: "alpha&beta", shouldError: false},
+		{name: "mixed operators", expression: "alpha&beta|gamma", shouldError: true},
```
What's wrong with mixed operators?

JoelSpeed commented on pkg/webhook/parser.go:
```diff
@@ -440,6 +453,98 @@ func (Generator) RegisterMarkers(into *markers.Registry) error {
 }
 
+// FeatureGateMap represents enabled feature gates as a map for efficient lookup
+type FeatureGateMap map[string]bool
```
The feature gate parsing logic appears to be needed in several places, should we create a centralised utility for this so that we only have to implement it once?

wazery replied:
yes totally agree, and adressed this

JoelSpeed commented on pkg/webhook/parser.go:
```diff
@@ -440,6 +453,98 @@ func (Generator) RegisterMarkers(into *markers.Registry) error {
+		}
+	}
+
+	// Check for mixing AND and OR operators
+	hasAnd := strings.Contains(expr, "&")
+	hasOr := strings.Contains(expr, "|")
+	if hasAnd && hasOr {
+		return fmt.Errorf("cannot mix '&' and '|' operators in feature gate expression: %s", expr)
+	}
```
I don't know if I've seen a use case for this, but in most logic I'd expect to be able to do `(A&B)|C` which would be equivalent to omitting the `()` since `OR ` takes precedence over `AND` in the ordering, right?

wazery replied:
thanks a lot for the valuable feedback here. I addressed this in the centralised feature gate util code, it should handle this logic now

JoelSpeed commented on pkg/webhook/parser.go:
```diff
@@ -440,6 +453,98 @@ func (Generator) RegisterMarkers(into *markers.Registry) error {
+	// Handle OR logic (any gate can be enabled)
+	if strings.Contains(featureGateExpr, "|") {
+		gates := strings.Split(featureGateExpr, "|")
+		for _, gate := range gates {
```
Are you familiar with the K8s sets utility?

If you were to turn the list of enabled feature gates into a set, I believe the `HasAny` and `HasAll` would simplify this a lot

Take a look at https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets

In the implementation I have downstream, we keep sets of All, Enabled and Disabled, so that every gate we find in a marker can be checked if we first know the state (It exists in the `All` set), and then if it's `Enabled` or not. Disabled doesn't come into this part of the use case thinking about it.

Might want something similar here?

wazery replied:
thanks a lot for the hint, I addressed this and used the provided package

JoelSpeed commented on pkg/webhook/parser.go:
```diff
@@ -440,6 +453,98 @@ func (Generator) RegisterMarkers(into *markers.Registry) error {
+	}
+
+	// Single gate logic
+	enabled, exists := enabledGates[featureGateExpr]
```
Either the AND or the OR logic would actually pass this through as well 🤔 

Another example of similar logic in https://github.com/openshift/kubernetes-sigs-controller-tools/blob/94812131024363d66cd1680f866bf694c560db39/pkg/crd/markers/patch_validation.go#L135-L147

In that example we have the `OR` options as `m.FeatureGateNames` and the `AND` options as `m.RequiredFeatureGateNames`
 
When there's only one gate to handle, it is a singleton in the `OR` set, but it could be in either set and still be functionally equivalent

JoelSpeed commented on pkg/webhook/parser.go:
```diff
@@ -440,6 +453,98 @@ func (Generator) RegisterMarkers(into *markers.Registry) error {
+	featureGateExpr := config.FeatureGate
+
+	// Handle AND logic (all gates must be enabled)
+	if strings.Contains(featureGateExpr, "&") {
```
This logic should also be centralised, perhaps we want the util to have centralised parsing and then a centralised `IsEnabledByFeatureGates` that would take whatever the `featureGate` value is from the marker/webhook/rbac and return a boolean to act upon?

wazery replied:
thanks a lot for this comment. It's addressed now

wazery commented on pkg/crd/gen.go:
```diff
@@ -32,6 +32,46 @@ import (
+func (fg FeatureGateMap) isEnabled(gateName string) bool {
+	enabled, exists := fg[gateName]
+	return exists && enabled
+}
```
this is addressed, thanks a lot for this valuable feedback 🙇

wazery committed:
feat: implement type-level feature gates for CRDs

Add +kubebuilder:featuregate marker support for conditional CRD type
generation with OR/AND expression support.

wazery committed:
wip

JoelSpeed commented:
> While rewriting the PR to have a centralised utility for supporting feature gates, I found it would require modifying each validation marker definition and the schema generation logic - a significant undertaking that should be its own focused effort. IMHO we can do that in another follow up PR.

I'm currently working with the declarative validation folks and the API review folks upstream to work out what feature gating might look like in terms of markers there.

I think we may want to copy their pattern downstream which is slightly more general, and uses chaining. For example, the `+k8s:ifEnabled(<option>)` marker must take a "payload". The payload in this case is another marker. So, rather than changing our existing markers, we would need to work out how to support `+k8s:ifEnabled(<option>):=+kubebuilder:validation:MinItems:=5`

The syntax and behaviour upstream is not quite cemented yet but I'm hoping that during the 1.35 cycle it will be, and that will allow us to start thinking about how this might work in the world of custom resources

JoelSpeed commented on pkg/crd/markers/featuregate.go:
```diff
@@ -0,0 +1,36 @@
+limitations under the License.
+*/
+
+package markers
```
What's the motivation for this being in a separate file?

JoelSpeed commented on pkg/crd/markers/crd.go:
```diff
@@ -57,6 +57,9 @@ var CRDMarkers = []*definitionWithHelp{
 	must(markers.MakeDefinition("kubebuilder:selectablefield", markers.DescribesType, SelectableField{})).
 		WithHelp(SelectableField{}.Help()),
+
+	must(markers.MakeDefinition("kubebuilder:featuregate", markers.DescribesType, TypeFeatureGate(""))).
+		WithHelp(TypeFeatureGate("").Help()),
```
Can you not register the same marker as both a Type and Field level marker without duplicating the marker as we currently have?

wazery committed:
feat: finalize feature gate implementation with latest changes
